### PR TITLE
Make JSON Decoder Continue After Error

### LINKED ISSUE #3392: Skip the wrong JSON line.
https://github.com/apache/arrow-rs/blob/17b3210af2ccd190489de9c641fd10f009abd45b/arrow-json/src/reader.rs#L243

If only the JSON data format in the middle line is incorrect, all the subsequent lines will fail to be converted. Can you provide a way to skip the wrong line?

### PR DESCRIPTION
**Pull request by yourenawo**
# Which issue does this PR close?
Closes #3392

### CODE REVIEW

**Review by tustvold — COMMENTED** [review]
I wonder if it might be a better UX to still return the error but continue to yield RecordBatch if the user keeps polling the iterator? This would let the user surface the error how they deem fit, e.g. log a warning, instead of silently dropping data?

I also think this needs some test coverage

**Comment by yourenawo** [comment]
sure, I try to add a callback function to solve this problem.

**Comment by tustvold** [comment]
Hmm... I'm not sure about this, what is the reason to prefer a callback over just yielding errors through the iterator trait?

I.e. Iterator::next might return

```
Some(Ok(batch1))
Some(Ok(batch2))
Some(Err(e))
Some(Ok(batch3_skipping_error))
Some(Ok(batch4))
None
```

**Comment by yourenawo** [comment]
ok, I think so, I added the parameter of recording error information to the decoder, so it needs to be mut

**Comment by tustvold** [comment]
I plan to have a play around with this this afternoon

File: arrow-json/src/reader.rs [hunk 1]
```diff
@@ -677,17 +689,29 @@ impl Decoder {
...
-        &self,
+        &mut self,
         value_iter: &mut I,
     ) -> Result<Option<RecordBatch>, ArrowError>
     where
         I: Iterator<Item = Result<Value, ArrowError>>,
     {
+        if let Some(ref s) = self.error_json {
+            let msg = s.clone();
+            self.error_json = None;
```
Comment (tustvold) [inline arrow-json/src/reader.rs:700]:
```suggestion
        if let Some(msg) = self.error_json.take() {
```

File: arrow-json/src/reader.rs [hunk 2]
```diff
@@ -699,8 +723,15 @@ impl Decoder {
             }
         }
         if rows.is_empty() {
-            // reached end of file
-            return Ok(None);
+            match self.error_json {
```
Comment (tustvold) [inline arrow-json/src/reader.rs:726]: Could use take here

Comment (tustvold) [inline arrow-json/src/reader.rs:726]: Could use take here

File: arrow-json/src/reader.rs [hunk 3]
```diff
@@ -677,17 +689,29 @@ impl Decoder {
     ///
     /// Returns `None` if the input iterator is exhausted.
     pub fn next_batch<I>(
-        &self,
+        &mut self,
```
Comment (tustvold) [inline arrow-json/src/reader.rs:692]: This is a breaking change but I support it, it opens the door to making Decoder a regular iterator

File: arrow-json/src/reader.rs [hunk 4]
```diff
@@ -677,17 +689,29 @@ impl Decoder {
...
+            return Err(ArrowError::JsonError(msg));
+        }
         let batch_size = self.options.batch_size;
         let mut rows: Vec<Value> = Vec::with_capacity(batch_size);
 
         for value in value_iter.by_ref().take(batch_size) {
-            let v = value?;
+            let v = match value {
+                Ok(v) => v,
+                Err(ArrowError::JsonError(s)) => {
```
Comment (tustvold) [inline arrow-json/src/reader.rs:709]: What do you think of handling all error variants the same way?

**Review by tustvold — COMMENTED** [review]
Sorry for being slow to review, just some minor nits and I think this is good to go 👍

**Comment by Goodjooy** [comment]
I am not understanding what this PR's change.

- if want handle as much as posiable row in a single batch (even though there might have one or many bad JSON record). returning the error messages togeter with the return value will better . like this : `Result<Option<(RecordBatch, Vec<ErrorSet>)>, ArrowError>` . Returning the error in last batch in `next_batch` is also make user confuse.


If just do not want loss any rows. becouse the `Iterator` of rust is lazy, once an error occurred the remain rows of this batch is still in the orginal Iterator that can be take in next batch. for a better understanding ,I am writting a rust-playgound [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5f77d983ffa876d2beeff2fcaa21aada)

**Comment by Goodjooy** [comment]
I think the error is returning with the return value is better.
The return type of the function signal can like : `Result<Option<(RecordBatch, Option<ArrowError>)>, ArrowError>` .
Returning previous error in `next_batch`  seems not a good idea

**Comment by tustvold** [comment]
What would you do if there are multiple errors

> Returning previous error in next_batch seems not a good idea

That isn't quite what this does, if there is an error it returns any records already processed without issue, and then returns the error

File: arrow-json/src/reader.rs (hunk 4 above)
Comment (tustvold) [inline arrow-json/src/reader.rs:709]: I think this also needs `if !values.is_empty()` to avoid returning an empty batch

Comment (yourenawo) [reply arrow-json/src/reader.rs:707]: ok, next rows have this

**Comment by tustvold** [comment]
Sorry for the delay with this, I'm currently reimplementing the JSON reader to not be row-based, and instead process batches at a time to drastically improve performance.

Whilst initially this will be a separate opt-in RawDecoder the eventual goal would be to replace the current implementation. I'm therefore a bit apprehensive about this feature, as it will likely be impossible to port across...

On a related note, such a behaviour is not possible for CSV for much the same reason.

This leads me to the position of not really knowing how best to proceed... Could you perhaps expand a bit on the use-case for this? Perhaps we could parse errors as nulls or something instead?

Sorry to contradict my earlier statements, but I'm wary of painting ourselves into a corner here

**Comment by yourenawo** [comment]
Because the source of json data is external, some dirty data will be introduced

**Comment by tustvold** [comment]
> some dirty data will be introduced

Could you expand on this, do you mean data with a mismatching schema, or data that is not even valid JSON? This PR appears to be intending to handle the latter case and I'm confused about how this would arise?

**Comment by yourenawo** [comment]
Invalid json data. When the program encounters an exception, a broken line json is written

**Comment by tustvold** [comment]
I'm afraid this approach is incompatible with the new vectorised RawReader, so closing. Apologies once again for the back and forth

**Pull request closed without merging**